src/tests/Microsoft.PowerFx.Interpreter.Tests/RecalcEngineTests.cs

```diff
@@ -239,6 +242,145 @@ public void FormulaCantRedefine()
+                    FormulaType.Number,
+                    new NamedFormulaType("x", FormulaType.Number)));
+            var result = recalcEngine.Eval("foo(0)");
+            Assert.Equal(2.0, result.ToObject());
+            Assert.False(enumerable.Any());
+        }
+
+        [Fact]
+        public void DefSimpleRecursiveFunc()
+        {
+            var config = new PowerFxConfig(null);
+            var recalcEngine = new RecalcEngine(config);
+            Assert.False(recalcEngine.DefineFunctions(
+                new UDFDefinition(
+                    "foo",
+                    "foo()",
+                    FormulaType.Blank)).Any());
+            var result = recalcEngine.Eval("foo()");
+            Assert.IsType<ErrorValue>(result);
+        }
+
+        [Fact]
+        public void DefHailstoneSequence()
+        {
+            var config = new PowerFxConfig(null);
+            var recalcEngine = new RecalcEngine(config);
+            var body = @"If(Not(x = 1), If(Mod(x, 2)=0, hailstone(x/2), hailstone(3*x+1)), x)";
+            var funcName = "hailstone";
+            var returnType = FormulaType.Number;
+            var variable = new NamedFormulaType("x", FormulaType.Number);
+
+            Assert.False(recalcEngine.DefineFunctions(
+                new UDFDefinition(funcName, body, returnType, variable)).Any());
+            Assert.Equal(1.0, recalcEngine.Eval("hailstone(192)").ToObject());
+        }
+
+        [Fact]
+        public void DefMutualRecursionFunc()
+        {
+            var config = new PowerFxConfig(null);
+            var recalcEngine = new RecalcEngine(config);
+            var bodyEven = @"If(number = 0, true, odd(Abs(number)-1))";
+            var bodyOdd = @"If(number = 0, false, even(Abs(number)-1))";
+
+            var udfOdd = new UDFDefinition(
+                "odd",
+                bodyOdd,
+                FormulaType.Boolean,
+                new NamedFormulaType("number", FormulaType.Number));
+            var udfEven = new UDFDefinition(
+                "even",
+                bodyEven,
+                FormulaType.Boolean,
+                new NamedFormulaType("number", FormulaType.Number));
+
+            Assert.False(recalcEngine.DefineFunctions(udfOdd, udfEven).Any());
+
+            Assert.Equal(true, recalcEngine.Eval("odd(17)").ToObject());
+            Assert.Equal(false, recalcEngine.Eval("even(17)").ToObject());
+        }
+
+        [Fact]
+        public void RedefinitionError()
+        {
+            var config = new PowerFxConfig(null);
+            var recalcEngine = new RecalcEngine(config);
+            Assert.True(recalcEngine.DefineFunctions(
+                new UDFDefinition("foo", "foo()", FormulaType.Blank),
+                new UDFDefinition("foo", "x+1", FormulaType.Number)).Any());
+            try
+            {
+                recalcEngine.Eval("foo()");
+                Assert.True(false);
+            }
+            catch (System.AggregateException)

```

Comment:
>AggregateException [](http://example.com/codeflow?start=26&length=18)

Same here, shouldn't this be either: Foo isn't defined, if we decided both were erroneous, or executing correctly with the 1st one?

Reply:
All the exceptions turn into System.AggregateException for some reason when catching them in [Fact] functions. If you know how to resolve this it would be much better to catch the specific exception.

Reply:
This is because eval is async under the hood and the `eval()` function is just blocking on the result, which also winds up wrapping exceptions (@MikeStall, I recall you had opinions on this)

For your tests, you'd want to call EvalAsync and catch a specific exception, although, in these cases, these shouldn't throw, they should be either errors on the original call to register, or a compile time error on Eval.

Reply:
Hmmm, I wonder why they throw an error then.

Reply:
Wait no this throwing an error makes sense. It originally throws an error on the Register, but then if you try to use it it *also* throws an error.

Reply:
To get rid of the AggregateException, make this function async and use await EvalAsync()  instead. 
Then you can even do Assert.ThrowsAsync<> to test for the exception.

Reply:
Fixed